CT-1568 trigger config state machine for awaiting responder and drafting states

stephenrichards opened the pull request.

aliuk2012 commented on app/views/case_attachments/destroy.js.erb:
```diff
@@ -0,0 +1 @@
+$("tr#case_attachment_<%= @attachment.id %>").remove();
```
What if theres only one file then you kind of have an orphaned table headings floating around.What about something like this?
```
<% if @case.attachments.count == 0 %>
  $("#uploaded-files").remove();
<% else %>
  $("tr#case_attachment_<%= @attachment.id %>").remove();
<% end%>
```

I dont particular like the usage of ID here and there is another way to pick up which element triggered the "remove" action and then use some javascript to move up the dom until it gets to the tr to remove it. But I'll let this one go for the time being. :-)

misaka replied:
I was thinking the same, about the ID. I see two options here, neither of which are really simple:

1. We use the attachments' key here. We don't currently constrain attachment's key tobe unique in the DB, but we could and probably should, especially now that this PR introduces the use of `find_or_initialize_by` when creating the attachment. The other issue may be how to encode the key as it'll have slashes ... and also, whether we want to remove the obfuscating SHA hashing of the id, speaking of which ...
2. We use the hash of the ID, just like we do for the S3 key, except that this hash is really weak obfuscation and doesn't, imo, give any substantial security. What we should do is generate a random ID, store that in the DB, and use that as the publicly visible ID.

@stephenrichards @aliuk2012 what do you think of these two options? I kind of feel like the weak hash of the ID we currently used needs to be changed anyway, maybe before we go live as otherwise it'll entail a potentially entail a data migration along with moving stuff around in S3.

misaka replied:
Also, ya, good point about not leaving an empty "Uploaded Files" section when there are no files left. I'll change that.

stephenrichards replied:
I'm of the opinion that there is nothing wrong with using the id here  - it is a staff tool after all and you have to be logged into to use it, and I wouldn't use the SHA in the s3 key either.  If we had gone live with the app, then I would leave the SHA S3 key as it is, but as we haven't it's a trivial job to remove it and change it to using the id .

misaka commented on app/controllers/cases_controller.rb:
```diff
@@ -152,8 +155,9 @@ def set_case
-      key:                   "#{@case.attachments_dir('responses')}/${filename}",
+      key:                   uploads_key,
       success_action_status: '201',
       acl:                   'public-read'
```
Yup, don't need these to be publically readable.

Commit:
CT-1568 adds tests for permitted events

Commit:
CT-1568 events for trigger cases awaiting responder config state machine

Commit:
CT-1568 WIP drafting events trigger case

Commit:
WIP

stephenrichards committed:
CT-1568 Fixed extend_for_pit and specs

stephenrichards committed:
CT-1568 Fixed policies for extend for pit

stephenrichards committed:
CT-1568 Rebase develop

stephenrichards committed:
CT-1568 Cocde reformatting

stephenrichards committed:
CT-1568 PR Review issues

stephenrichards committed:
CT-1568 fix spec with new method signature

misaka committed:
CT-1568 add CasesController#remove_clearance for clarity

This action doesn't need to do anything, but it'd be nice if it was there, so
that it's less confusing when trying to understand the system, and if it
authorized to ensure that users don't even think they can remove clearance if
they happen to get to load this page somehow.

misaka committed:
CT-1568 adapt unflag_for_clearance to new state machine

Both the controller and the state machine used to call the policy
BasePolicy#can_unflag_for_clearance?. This moves the policy logic into
Workflows::Predicates, points the state machine to use that, and changes the
policy to check whether the state machine can trigger the event.

misaka committed:
CT-1568 add disclosure_specialist_coworker to predicate specs

misaka committed:
CT-1568 add missing events to permitted_events for approvers

misaka committed:
CT-1568 disable Rubocop method length check for trigger_event

This method should be re-factored to extract code common to can_trigger_event?()
and possibly event_present_and_triggerable?().
